lib/fog/google/shared.rb

```diff
@@ -84,41 +65,31 @@ def initialize_google_client(options)
         end
 
         auth = nil
-        if options[:google_json_key_location] || options[:google_json_key_string]
-          if options[:google_json_key_location]
-            json_key_location = File.expand_path(options[:google_json_key_location])
-            json_key = File.open(json_key_location, "r", &:read)
-          else
-            json_key = options[:google_json_key_string]
-          end
 
-          json_key_hash = Fog::JSON.decode(json_key)
-          unless json_key_hash.key?("client_email") || json_key_hash.key?("private_key")
-            raise ArgumentError.new("Invalid Google JSON key")
-          end
-
-          options[:google_client_email] = json_key_hash["client_email"]
-          unless options[:google_client_email]
-            raise ArgumentError.new("Missing required arguments: google_client_email")
-          end
-
-          auth = ::Google::Auth::ServiceAccountCredentials.make_creds(
-            :json_key_io => StringIO.new(json_key_hash.to_json),
-            :scope => options[:google_api_scope_url]
-          )
+        if options[:google_json_key_location] || options[:google_json_key_string]
+          auth = process_key_auth(options)
         elsif options[:google_auth]
           auth = options[:google_auth]
+        elsif options[:application_default]
+          auth = process_application_default_auth(options)
         else
-          # fall back to Application Default Credentials before erroring
-          auth = ::Google::Auth.get_application_default(
-            options[:google_api_scope_url]
+          Fog::Logger.warning(
+              "Didn't detect any explicit auth settings, " \
+              "trying to use application default credentials."
+          )
+          auth = process_application_default_auth(options)
+        end
+
+        if auth.nil?
+          raise Fog::Errors::Error.new(
+              "Failed to configure authentication for Fog client.\n" \
+                "Check your auth options, must be one of:\n" \
+                "- :google_json_key_location,\n" \
+                "- :google_json_key_string,\n" \
+                "- :google_auth,\n" \
```

Comment:
This is a bit misleading in the the case when a user explicitly sets `:application_default` and doesn't have ADC properly configured because `process_application_default_auth()` will return nil and the user will see an error message that fog is misconfigured when the reality is that the need to debug their ADC.

Reply:
Agreed. Fixed - now fallback auth has a separate method and is not rescuing exceptions if the auth config is explicit.